system/src/system_network_internal.h

```diff
@@ -599,34 +613,49 @@ class ManagedNetworkInterface : public NetworkInterface
         return (SPARK_WLAN_STARTED && WLAN_CONNECTING);
     }
 
-    void on() override
+    int on() override
     {
         LOG_NETWORK_STATE();
-        if (!SPARK_WLAN_STARTED)
+        if (!SPARK_WLAN_STARTED || !WLAN_INITIALIZED)
         {
+            bool startedPowerOn = !SPARK_WLAN_STARTED;
+            WLAN_INITIALIZED = 0;
             const auto diag = NetworkDiagnostics::instance();
             diag->status(NetworkDiagnostics::TURNING_ON);
-            system_notify_event(network_status, network_status_powering_on);
+            if (startedPowerOn) {
+                // Suppress these events
+                system_notify_event(network_status, network_status_powering_on);
+            }
             config_clear();
             const int ret = on_now();
             if (ret != 0) {
                 diag->lastError(ret);
+                if (!wlan_watchdog_duration) {
+                    // Just in case arm WLAN WD
+                    ARM_WLAN_WD(POWER_ON_WD);
+                }
+            } else {
+                WLAN_INITIALIZED = 1;
```

Comment:
If we get past this block and have failed to turn on the NCP `if (ret != 0)`, shouldn't we not be changing the LED to breathing blue yet?  Should that also be moved into the `if (!ret) { }` block?

Reply:
Again, I don't want to introduce any UI changes so late into 2.x.

Reply:
But you moved other code into that new block which changes behavior of system events and diagnostics :D. Also, we are fixing bugs ... not changing the way the UI is interpreted.  Gen 2 breathing white was always booting, breathing blue is NCP on, blinking green network connecting, etc..  If nothing else though, we should add a FIXME note.

Reply:
The behavior of system events and diagnostics is unchanged. The changes that had to be made are exactly so that we don't generate a torrent of events if we are waiting for power-on sequence to succeed.

Reply:
Before if cellular_on() failed, it would have still generated the `network_status_on` event... now it waits until cellular is really on, right?

Reply:
Hm. Perhaps that `if (!ret)` should instead be `if (startedPowerOn)` then, right?

Reply:
maybe `if (WLAN_INITIALIZED)` ?  don't we want to generate `network_status_on` after it's fully on, and not just in the process still of `network_status_powering_on` ?

Reply:
if WLAN_INITIALIZED is the same as if !ret though.

Reply:
Before it would just fail and say it succeeded via system LED and events.  Now we have an opportunity to fix that incorrect behavior.

Reply:
We can both agree on the fact that this needs to be fixed, but:

1. Something out there might be relying on this broken behavior (events)
2. The LED indication might be confusing as we might get stuck in breathing white for an extended period of time (up to several minutes)

Reply:
We could open a CH ticket regarding this, and reference it in a FIXME comment in this code and move on :D But you'll also want to take the code out of the `if (!ret)` block.  I also don't think being stuck in breathing blue or blinking green is any less confusing if the real issue is the NCP hasn't finished turning on.

Reply:
I'll make all the other changes and will play around with this part to see how it's going to behave, so let's have a separate discussion tomorrow and come up with some sort of a decision.